clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1614,73 +1616,63 @@ mlir::Value ScalarExprEmitter::VisitUnaryLNot(const UnaryOperator *E) {
   llvm_unreachable("destination type for negation unary operator is NYI");
 }
 
+// Conversion from bool, integral, or floating-point to integral or
+// floating-point. Conversions involving other types are handled elsewhere.
+// Conversion to bool is handled elsewhere because that's a comparison against
+// zero, not a simple cast.
 mlir::Value ScalarExprEmitter::buildScalarCast(
     mlir::Value Src, QualType SrcType, QualType DstType, mlir::Type SrcTy,
     mlir::Type DstTy, ScalarConversionOpts Opts) {
-  // The Element types determine the type of cast to perform.
-  mlir::Type SrcElementTy;
-  mlir::Type DstElementTy;
-  QualType SrcElementType;
-  QualType DstElementType;
-  if (SrcType->isMatrixType() && DstType->isMatrixType()) {
-    llvm_unreachable("NYI");
-  } else {
-    assert(!SrcType->isMatrixType() && !DstType->isMatrixType() &&
-           "cannot cast between matrix and non-matrix types");
-    SrcElementTy = SrcTy;
-    DstElementTy = DstTy;
-    SrcElementType = SrcType;
-    DstElementType = DstType;
-  }
```

Comment:
Awesome. If `buildScalarCast()` gets called by some other user at some point we won't catch it here, perhaps just add one more extra assert `assert(!SrcType->isMatrixType() && !DstType->isMatrixType() && "matrix types not supposed to be handled by this function")` ?

Reply:
Done.